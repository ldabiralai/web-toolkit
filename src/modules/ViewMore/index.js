import React from 'react';
import PropTypes from 'prop-types';
import styled, { css } from 'react-emotion';
import { rgba } from 'polished';
import { coreLightMinus1, coreDarkPlus1 } from '../../colors';
import { fontFamilies } from '../../typography';

const StyledButton = styled.span`
  color: ${coreLightMinus1};
  font-family: ${fontFamilies.alphaHeadline};
  border: 1px solid ${rgba(coreLightMinus1, 0.35)};
  background: ${rgba(coreDarkPlus1, 0.2)};
  border-radius: 3px;
  text-transform: uppercase;
  font-size: 11px;
  padding: 6px 14px;
  line-height: 11px;
  &:hover {
    cursor: pointer;
  }
`;

const StyledList = styled.ul`
  ${props =>
    props.blockHeight !== 0 &&
    css`
      max-height: ${props.blockHeight}px;
    `}
  overflow: hidden;
  transition: max-height 400ms ease;
`;

const StyledItem = styled.li`
  margin: 0;
  overflow: hidden; /* prevent collapsing margins */
`;

export default class ViewMore extends React.Component {
  listRef = React.createRef();

  transitionEndCallback = this.transitionEndCallback.bind(this);

  state = {
    itemsToShow: 1,
    expanded: false,
    height: 0,
  };

  componentDidMount() {
    const { children } = this.props;
    if (!children || !children.length) return;
    this.setState({ height: this.getItemHeight() });
    this.listRef.current.removeEventListener('transitionend', this.transitionEndCallback);
    this.listRef.current.addEventListener('transitionend', this.transitionEndCallback);
  }

  getItemHeight() {
    return this.listRef.current.children[0].clientHeight;
  }

  transitionEndCallback() {
    const { expanded } = this.state;
    if (!expanded) {
      this.setState({ expanded: true });
    } else {
      this.setState({ itemsToShow: 1, expanded: false });
    }
  }

  showMore() {
    const { children } = this.props;
    this.setState({ height: this.getItemHeight() * children.length, itemsToShow: children.length });
  }

  showLess() {
    this.setState({ height: this.getItemHeight() });
  }

  handleClick() {
    const { expanded } = this.state;
    if (!expanded) {
      this.showMore();
    } else {
      this.showLess();
    }
  }

  render() {
    const { children } = this.props;
    const { expanded, itemsToShow, height } = this.state;
    if (!children || !children.length) return null;
    return (
      <>
        <StyledList isExpanded={expanded} innerRef={this.listRef} blockHeight={height}>
          {children.slice(0, itemsToShow).map(child => (
            <StyledItem key={child.key}>{child}</StyledItem>
          ))}
        </StyledList>
        {children.length > 1 && (
          <StyledButton onClick={() => this.handleClick()}>{expanded ? 'Show less' : 'Show more'}</StyledButton>
        )}
      </>
    );
  }
}

ViewMore.propTypes = {
  children: PropTypes.arrayOf(PropTypes.node).isRequired,
};
