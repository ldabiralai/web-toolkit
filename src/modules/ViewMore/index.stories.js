import React from 'react';
import { storiesOf } from '@storybook/react';
import { withInfo } from '@storybook/addon-info';
import styled from 'react-emotion';
import { object } from '@storybook/addon-knobs';
import { ViewMore } from '../..';
import { coreLightMinus1 } from '../../colors';

const viewmore = storiesOf('Modules|ViewMore', module).addDecorator(withInfo);

const Wrapper = styled.div`
  color: ${coreLightMinus1};
`;

const Element = styled.div`
  margin-bottom: 15px;
`;

const children = [];
for (let i = 0; i < 10; i += 1) {
  children.push(<Element key={i}>match</Element>);
}

viewmore.add('viewmore', () => (
  <Wrapper>
    <ViewMore>{object('children', children)}</ViewMore>
  </Wrapper>
));
