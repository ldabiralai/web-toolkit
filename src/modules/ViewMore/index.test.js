import React from 'react';
import { mount } from 'enzyme';
import { object } from '@storybook/addon-knobs';
import ViewMore, { StyledButton } from '.';

const children = [];
for (let i = 0; i < 10; i += 1) {
  children.push(
    <div data-test="children" key={i}>
      match
    </div>
  );
}

describe('ViewMore', () => {
  it('renders its content', () => {
    expect(mount(<ViewMore>{object('children', children)}</ViewMore>)).toMatchSnapshot();
  });
  describe('items behavior', () => {
    it('should display one item at first load', () => {
      const wrapper = mount(<ViewMore>{object('children', children)}</ViewMore>);
      expect(wrapper.find('[data-test="children"]').length).toEqual(1);
    });
    it('should display more items when view more button is clicked', () => {
      const wrapper = mount(<ViewMore>{object('children', children)}</ViewMore>);
      wrapper.find(StyledButton).simulate('click');
      expect(wrapper.find('[data-test="children"]').length).toEqual(10);
    });
    it('should display one item when view less button is clicked', async () => {
      const wrapper = mount(<ViewMore>{object('children', children)}</ViewMore>);
      wrapper.find(StyledButton).simulate('click');
      wrapper.find('ul').simulate('transitionEnd');
      wrapper.find(StyledButton).simulate('click');
      expect(wrapper.find('[data-test="children"]').length).toEqual(1);
    });
  });
});
