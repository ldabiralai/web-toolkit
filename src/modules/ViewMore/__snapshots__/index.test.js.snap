// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ViewMore renders its content 1`] = `
<ViewMore>
  <Styled(ul)
    blockHeight={0}
    data-test="item-list"
    innerRef={
      Object {
        "current": <ul
          class="css-n93gab e1e87viu1"
          data-test="item-list"
        >
          <li
            class="css-1u41msn e1e87viu2"
          >
            <div
              data-test="children"
            >
              match
            </div>
          </li>
        </ul>,
      }
    }
    isExpanded={false}
  >
    <ul
      className="css-n93gab e1e87viu1"
      data-test="item-list"
    >
      <Styled(li)
        key="0"
      >
        <li
          className="css-1u41msn e1e87viu2"
        >
          <div
            data-test="children"
            key="0"
          >
            match
          </div>
        </li>
      </Styled(li)>
    </ul>
  </Styled(ul)>
  <Styled(span)
    data-test="view-button"
    onClick={[Function]}
  >
    <span
      className="css-41z9q8 e1e87viu0"
      data-test="view-button"
      onClick={[Function]}
    >
      Show more
    </span>
  </Styled(span)>
</ViewMore>
`;
